File: leetcode/DFS/combination Sum.go
package DFS


func combinationSum(candidates []int, target int) [][]int {
	res := [][]int {}
	each := []int{}
	sort.Ints(candidates)
	dfs(candidates, 0, target, each, &res)
	return res
}

func dfs(candidates []int, start int,  target int, each []int, res *[][]int ){
	if target == 0{
		copy := append([]int{}, each...)
		*res = append(*res, copy)
		return
	}
	for i := start; i < len(candidates); i++{
		if target < candidates[i]{
			return
		}
		each = append(each, candidates[i])
		dfs(candidates, i, target-candidates[i], each, res)
		each = each[0:len(each)-1]
	}
}

File: leetcode/archived/C++/Combination_sum.cpp
class Solution {
	public:
		vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
			vector<vector<int> >result;
			vector<int> output;
			sort(candidates.begin(), candidates.end());
			search(result, output, target, candidates.begin(), candidates.end());

			return result;
		}

		void search(vector<vector<int> >&result, vector<int>output, int target, vector<int>::iterator begin, vector<int>::iterator end)
		{
			if(0 == target)
			{
				result.push_back(output);
				return;
			}

			if(*begin <= target)
			{
				output.push_back(*begin);
				search(result, output, target-*begin, begin, end);
				output.pop_back();
			}

			if(begin+1 < end)
				search(result, output, target, begin+1, end);
		}
};
